### PR TITLE
fix(arel): PR 31 — Cte boolean tristate + BindError JSON.stringify(sql)

### PR DESCRIPTION
**Pull request by deanmarano**
Part of `docs/arel-alignment-followup-prs.md`, PR 31.

## Changes

**1. `Cte.materialized` boolean tristate**

Rails `cte.rb` stores `materialized` as `true | false | nil`. The old TS type was `"materialized" | "not_materialized"` strings. Widened to `boolean | null`. Visitor now branches on `=== true` / `=== false`, matching `to_sql.rb:736`. `null`/`undefined` emits no modifier. Updated 4 call sites in tests.

**2. `BoundSqlLiteral` construction-time validation**

Already implemented in a prior PR — `validate()` is called in the constructor. No changes needed.

**3. `BindError` message uses `JSON.stringify(sql)`**

Rails `errors.rb:14`: `"#{message} in: #{sql.inspect}"` where `.inspect` quotes and escapes the string. The TS equivalent is `JSON.stringify(sql)`. Fixed `errors.ts` and updated two expected messages in `bound-sql-literal.test.ts`.

## Verification

- `pnpm test` — all pass
- `pnpm api:compare --package arel` — 884/884

### CODE REVIEW

File: packages/arel/src/nodes/cte.ts [hunk 1]
```diff
@@ -10,9 +10,9 @@ import { Table } from "../table.js";
 export class Cte extends Binary {
   readonly name: string;
   readonly relation: Node;
-  readonly materialized?: "materialized" | "not_materialized";
+  readonly materialized?: boolean | null;
 
-  constructor(name: string, relation: Node, materialized?: "materialized" | "not_materialized") {
+  constructor(name: string, relation: Node, materialized?: boolean | null) {
```
[the remark this answers is not in the document]
Comment (deanmarano) [reply packages/arel/src/nodes/cte.ts]: This is an intentional pre-release migration. Rails' `Cte#initialize` takes `materialized:` as a keyword arg accepting `true`, `false`, or `nil` (the default). The previous TS string enum (`"materialized" | "not_materialized"`) was a non-Rails deviation. Since the package is pre-release with no published external consumers, a clean type swap is the right call — a normalization shim would carry the old non-Rails shape indefinitely. All in-tree call sites have been updated.

File: packages/arel/src/errors.ts [hunk 2]
```diff
@@ -14,7 +14,7 @@ export class EmptyJoinError extends ArelError {
 
 export class BindError extends ArelError {
   constructor(message: string, sql?: string) {
-    super(sql ? `${message} in: ${sql}` : message);
+    super(sql !== undefined ? `${message} in: ${JSON.stringify(sql)}` : message);
```
[the remark this answers is not in the document]
Comment (deanmarano) [reply packages/arel/src/errors.ts:17]: Acknowledged — `JSON.stringify` differs from Ruby's `String#inspect` for control characters. This is the accepted TS analog documented in the spec ("closest TS analog to Ruby `.inspect`"). SQL containing raw control characters in practice is extremely rare, and the important cases (embedded quotes, backslashes) are correctly handled. If a closer approximation is ever needed, a dedicated `rubyInspect()` helper can be swapped in without changing the `BindError` API.

File: packages/arel/src/nodes/bound-sql-literal.test.ts [hunk 3]
```diff
@@ -52,7 +52,15 @@ describe("BoundSqlLiteralTest", () => {
...
+        'wrong number of bind variables (2 for 3) in: "id IN (?, ?, ?)"',
+      );
+    });
+
+    it("JSON.stringify escapes embedded double-quotes in error message", () => {
+      // SQL has 1 placeholder; 2 binds triggers a count mismatch.
+      // Verifies JSON.stringify correctly escapes the embedded double-quote in the SQL.
+      expect(() => new Nodes.BoundSqlLiteral('name = "O\'Brien" AND ?', [1, 2])).toThrow(
+        'wrong number of bind variables (2 for 1) in: "name = \\"O\'Brien\\" AND ?"',
       );
```
[the remark this answers is not in the document]
Comment (deanmarano) [reply packages/arel/src/nodes/bound-sql-literal.test.ts:70]: The test intentionally encodes `JSON.stringify`'s current escaping behavior. If the implementation is ever updated to a closer `String#inspect` approximation, the test will need to be updated too — that's by design. The test documents what the code actually does, not what Ruby does, so a divergence will surface as a failing test prompting a deliberate update rather than a silent regression.

**Review by deanmarano — COMMENTED** [review]
Both comments ask us to diverge from Rails — the current behavior is correct parity.

Rails `bound_sql_literal.rb:9-10`:
```ruby
has_positional = !(positional_binds.nil? || positional_binds.empty?)
has_named      = !(named_binds.nil?      || named_binds.empty?)
```
Named-bind validation only runs inside `elsif has_named` (line 19), so when `named_binds` is `{}` (or `nil`), validation is skipped entirely. `BoundSqlLiteral.new("id = :foo", [], {})` does **not** raise in Rails — the `:foo` token is simply left in the SQL for the visitor to handle. Our `validate()` guard `Object.keys(namedBinds).length > 0` is a direct translation of `has_named`.

The audit doc entry is therefore correctly marked resolved: we validate at construction time whenever binds are actually supplied, matching Rails exactly. Adding a test for the empty-binds case would pin behaviour that Rails itself does not enforce, and adding a test that "it raises when `{}` is passed" would be wrong — Rails does not raise there.
